Add the "writing a good package page" page

jayoung-lee opened the pull request:
Staged: https://dartlang-org-staging-0.firebaseapp.com/guides/libraries/writing-package-pages

parlough commented on src/_guides/libraries/writing-package-pages.md:
```diff
@@ -0,0 +1,192 @@
+
+Most package users who landed on your package on pub.dev will quickly scan your README to decide whether to try your package or not. A nice README will catch users’ attention and show that your package is worth trying.
+
+There are seven tips that you can follow to write a good README. 
```
I'm curious if it is worth mentioning that it may be better to not include a Markdown title in the README as otherwise it will appear multiple times on pub.dev. Many of the Dart team's packages previously removed theirs as a result of this. An example of a package that still does include a Markdown title though is [dartdoc](https://pub.dev/packages/dartdoc). There is an [pub.dev issue](https://github.com/dart-lang/pub-dev/issues/3238) tracking automatic removal of these, but it seems to be on hold.

![image](https://user-images.githubusercontent.com/18372958/114804178-8182f380-9d66-11eb-9780-e7c1a49693e0.png)

kwalrath replied:
This is mentioned at the end of the first paragraph of tip 1, but it's easy to miss. I'll think about what to do...

parlough commented on src/_guides/libraries/writing-package-pages.md:
```diff
@@ -0,0 +1,192 @@
+
+### 2. Include visual content to explain things
+
+UI is difficult to explain in text. If your package page is just a “wall of text” with no visual content, users may find it intimidating. In that case, it is very likely that package searchers skip your package without reading the README. Place visual content such as images, gifs, movs close to the beginning of the README, so that users can easily find them and use them. (Movs or gifs are even be ...
```
Perhaps a generic term like "video" rather than "mov" would be better, since more people will likely understand and I believe at least GitHub supports other container formats besides mov. Both are fine though :)

kwalrath commented:
I've just staged this: https://dartlang-org-staging-0.firebaseapp.com/guides/libraries/writing-package-pages

kwalrath commented:
@jayoung-lee could you please produce a side-by-side figure for https://dartlang-org-staging-0.firebaseapp.com/guides/libraries/writing-package-pages#2-include-visual-content-to-explain-things? Currently the two screenshots show up one after the other (and huge!), but the text talks about the left and right figures (which would be better).

jayoung-lee commented:
I added a side-by-side figure and deleted the old ones. Hope this works!

Is there a way to make all the images to be like 70-80% of the page width? Other images look very big too.

kwalrath commented:
> I added a side-by-side figure and deleted the old ones. Hope this works!

Yes, much better!
 
> Is there a way to make all the images to be like 70-80% of the page width? Other images look very big too.

I'll try a few things and let you know how it goes.

kwalrath commented:
I've staged the new image styling: https://dartlang-org-staging-0.firebaseapp.com/guides/libraries/writing-package-pages

kwalrath commented on src/_guides/libraries/writing-package-pages.md:
```diff
@@ -0,0 +1,293 @@
+construct a good package page on pub.dev.
+More specifically, this document provides several tips on
+how to write a better README.md file,
+as marked in the following image.
```
Can you redo this figure? Two reasons:

* It doesn't include the title **url_launcher** in the area of **This document**, but that text _is_ provided by the README.
* It shows the title **url_launcher**. 😆  We probably want to show a better example, no?

Could we use in_app_purchase, yaml, or another nicely done package page?

kwalrath replied:
I wonder if we should change "This document" to "README file (this document)"

jayoung-lee replied:
The label "this document" is updated. And I replaced the image with `shared_preferences` page, which has a title `Shared preferences plugin`. Would this work?

Otherwise, we'll have to update one of the package's readme and retry. All the 1P Flutter Favorite packages have some sort of titles in the body, ironically :P

kwalrath replied:
I'll take another look Monday

kwalrath commented on src/_guides/libraries/writing-package-pages.md:
```diff
@@ -0,0 +1,293 @@
+
+{% asset libraries/package-page-sections.png alt="package page contains sections like package layout, flutter favorite, package scoring, verified publishers, pubspec file" class="screenshot" %}
+
+1. About the [package layout ](https://dart.dev/tools/pub/package-layout)
+2. About the [Flutter Favorite program](https://flutter.dev/docs/development/packages-and-plugins/favorites)
+3. About the [package scoring system](https://pub.dev/help/scoring)
+4. About [verified publishers](https://dart.dev/tools/pub/verified-publishers)
+5. About the [pubspec file](https://dart.dev/tools/pub/pubspec)
```
Let's put the description down here, closer to the links, and make the link text match the figure labels exactly:

```suggestion
For details about other parts of the package page,
follow these links:

1. [Package layout ](https://dart.dev/tools/pub/package-layout)
2. [Flutter Favorite](https://flutter.dev/docs/development/packages-and-plugins/favorites)
3. [Package scoring](https://pub.dev/help/scoring)
4. [Verified publishers](https://dart.dev/tools/pub/verified-publishers)
5. [Pubspec file](https://dart.dev/tools/pub/pubspec)
```

jayoung-lee replied:
Yes! Making commits sounds good to me. Thank you :D

kwalrath reviewed (commented):
I'm going to have a bunch of small copy edits. @jayoung-lee are you OK with me just making commits, or would you like to review the changes?

kwalrath commented on src/_guides/libraries/writing-package-pages.md:
```diff
@@ -0,0 +1,294 @@
+Place visual content such as images, gifs, movs
+close to the beginning of the README,
+so that users can easily find them and use them.
+(Movs or gifs are even better than static images!) 
```
Why are they better? Because they grab users' attention? Because users prefer them? Because they can convey more? (I'd like to be more specific here.)

Also, might you use an MP4 instead of a MOV?

kwalrath replied:
I'm thinking of saying something like this:

>  Tip: Prefer animated GIFs and videos, when you have a choice. They can be more effective than static images for attracting readers and conveying information.

jayoung-lee replied:
I like the new phrasing! 

Yeah, I'd say that UIs are not static in most cases (they're animated or they respond to user input), so videos or gifs are better in those cases, in terms of conveying information about the package. MP4 works too. :)

kwalrath commented:
@jayoung-lee I'm working on this today... and will have questions. :)

kwalrath commented on src/_guides/libraries/writing-package-pages.md:
```diff
@@ -0,0 +1,294 @@
+
+### 2. Include visual content to explain things
+
+UI is difficult to explain in text.
```
This makes it seem like visual content is only useful for UI-related plugins. I'll think about how to fix it...

kwalrath commented on src/_guides/libraries/writing-package-pages.md:
```diff
@@ -0,0 +1,294 @@
+
+Linking specific functions or classes to those in dartdoc can be useful too.
+See [async](https://pub.dev/packages/async) package for
+an example of linking dartdocs.
```
Yikes, these links aren't working!

jayoung-lee replied:
@jonasfj Do you know when the new version is going to be published?

kwalrath replied:
Filed https://github.com/dart-lang/async/issues/168

jonasfj replied:
I think we just need to publish it...

parlough commented on src/_guides/libraries/writing-package-pages.md:
```diff
@@ -38,130 +38,152 @@ follow these links:
-If there is important information users should know about,
-such as project status or constraints, note that at the top too. 
+*   `A Flutter plugin for showing rainbows.`
+*   `Use machine learning to categorize bird sounds.`
```
It's a plane...it's super man...it's a round hummingbird...it's dash!

kwalrath replied:
It's Mega-Dash! 😆

parlough commented on src/_guides/libraries/writing-package-pages.md:
```diff
@@ -0,0 +1,335 @@
+{{site.alert.end}}
+
+The screenshots below show how
+adding visual content made the `in_app_purchase` package page look better.
```
Maybe instead of saying it looks `better`, we could be a bit more specific. Something more targetted such as `more informative`, `more fleshed out`, `more engaging`, or similar (I'm not really sure) to better indicate how it is "better".

jayoung-lee replied:
What about "look informative at first glance"? :)

kwalrath commented:
I'm done with my copy edit, although I haven't yet addressed all of @parlough feedback. @jayoung-lee could you please take a look at my changes and at Parker's remaining comments? I've updated the staging server:

https://dartlang-org-staging-0.firebaseapp.com/guides/libraries/writing-package-pages

jayoung-lee commented:
LGTM! 

I made just one commit https://github.com/dart-lang/site-www/pull/3147/commits/d45ed4af9344e4714f3fdfaee73c9d485491b54d to address the comment from @parlough. I'll remove the `[PENDING]...` line, once https://github.com/dart-lang/async/issues/168 is closed.

Let me know if I missed anything.

kwalrath commented:
LGTM once we can remove that PENDING comment!

parlough commented:
@kwalrath @jayoung-lee Looks like the links in the async README are working now with the latest publish: https://pub.dev/documentation/async/latest/

kwalrath commented:
LGTM. We'll want to add links to this from other docs and from the sidenav, but that could be in a follow-up PR. I'm thinking that https://dart.dev/tools/pub/publishing and https://dart.dev/tools/pub/package-layout (at least) should link to this page.

As for the sidenav, maybe **Writing package pages** could go just above **Package reference**, under **Publishing packages**?

![image](https://user-images.githubusercontent.com/2164483/116620688-063d4800-a8f7-11eb-972f-1b2dcbd5c630.png)

jayoung-lee commented:
> We'll want to add links to this from other docs and from the sidenav, but that could be in a follow-up PR. I'm thinking that https://dart.dev/tools/pub/publishing and https://dart.dev/tools/pub/package-layout (at least) should link to this page.
> 
> As for the sidenav, maybe **Writing package pages** could go just above **Package reference**, under **Publishing packages**?

Both sound good to me :)
